File: js/create-prp.js
// ── CreatePRP — in-browser port of create_prp.py ──────────────────────────
// Merges the 3 raw exports (risk-export / assessment-export / tprm|supplier)
// into the consolidated PRP workbook the report pipeline consumes.
//
// Intentional deviations from create_prp.py:
//  · TPRM sheet is named 'TPRM Web-Portal Export' (the Python writes
//    'TPRM WebApp Portal', but reports d1/s1c and the shipped sample output
//    read 'TPRM Web-Portal Export' — the Python name appears to be drift).
//  · Every appended formula cell also carries its computed result
//    ({formula, result}) so the in-browser reports — which read cached
//    values, never recalculate — see real data. Excel still recalculates on
//    open (calcProperties.fullCalcOnLoad), keeping TODAY() fresh.
window.CreatePRP = (function () {
  'use strict';

  var SHEET_RISK = 'OneTrust - Risk Export';
  var SHEET_TPRM = 'TPRM Web-Portal Export';   // deviation — see header comment
  var SHEET_ASSESSMENT = 'OneTrust Assessment';
  var OUTPUT_NAME = 'PRP Sample Jun (2).xlsx';

  var MS_PER_DAY = 86400000;
  var EXCEL_EPOCH = Date.UTC(1899, 11, 30);    // Excel serial 0
  var SEP26_SERIAL = serialFromDate(new Date(Date.UTC(2026, 8, 30))); // DATE(2026,9,30)

  function _yield() { return new Promise(function (res) { setTimeout(res, 0); }); }

  // ── classify — mirrors find_source_files()'s filename checks ────────────
  function classify(fileName) {
    var lower = String(fileName || '').toLowerCase();
    if (!/\.xlsx$/.test(lower)) return null;
    if (lower.indexOf('risk-export') !== -1) return 'risk';
    if (lower.indexOf('assessment-export') !== -1) return 'assessment';
    if (lower.indexOf('tprm') !== -1 || lower.indexOf('supplier') !== -1) return 'tprm';
    return null;
  }

  // ── date parsing — the 5 strptime formats, in the Python's order ────────
  // %Y is exactly 4 digits; day/month accept 1-2; invalid calendar dates
  // (31/02/…) fall through to the next format, like strptime raising.
  var DATE_FORMATS = [
    { re: /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/, d: 1, m: 2, y: 3 },  // %d/%m/%Y
    { re: /^(\d{1,2})-(\d{1,2})-(\d{4})$/,  d: 1, m: 2, y: 3 },   // %d-%m-%Y
    { re: /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/, d: 2, m: 1, y: 3 },  // %m/%d/%Y
    { re: /^(\d{4})-(\d{1,2})-(\d{1,2})$/,  d: 3, m: 2, y: 1 },   // %Y-%m-%d
    { re: /^(\d{1,2})-(\d{1,2})-(\d{4})$/,  d: 2, m: 1, y: 3 },   // %m-%d-%Y
  ];
  // UTC construction is load-bearing: ExcelJS serializes Dates via UTC epoch
  // math, so a local-midnight Date in a UTC+ timezone saves as the previous day.
  function parseDateString(s) {
    var str = String(s).trim();
    for (var i = 0; i < DATE_FORMATS.length; i++) {
      var f = DATE_FORMATS[i], m = f.re.exec(str);
      if (!m) continue;
      var y = +m[f.y], mo = +m[f.m], d = +m[f.d];
      if (mo < 1 || mo > 12 || d < 1 || d > 31) continue;
      var dt = new Date(Date.UTC(y, mo - 1, d));
      if (dt.getUTCFullYear() !== y || dt.getUTCMonth() !== mo - 1 || dt.getUTCDate() !== d) continue;
      return dt;
    }
    return null;
  }

  function serialFromDate(d) { return Math.round((d.getTime() - EXCEL_EPOCH) / MS_PER_DAY); }
  function todayUTC() {
    var now = new Date();
    return new Date(Date.UTC(now.getFullYear(), now.getMonth(), now.getDate()));
  }

  // NETWORKDAYS(a,b): inclusive Mon–Fri count; negative when a > b.
  function networkdaysSerial(a, b) {
    var sign = 1;
    if (a > b) { var t = a; a = b; b = t; sign = -1; }
    var days = b - a + 1;
    var fullWeeks = Math.floor(days / 7);
    var count = fullWeeks * 5;
    var rem = days - fullWeeks * 7;
    var dow = new Date(EXCEL_EPOCH + a * MS_PER_DAY).getUTCDay(); // 0=Sun … 6=Sat
    for (var i = 0; i < rem; i++) {
      var d = (dow + i) % 7;
      if (d !== 0 && d !== 6) count++;
    }
    return sign * count;
  }

  // Excel-comparison view of a copied cell value:
  //  blank → serial 0 (Excel treats blank as 0 in arithmetic/comparisons),
  //  Date/number → its serial, anything else → text (arithmetic would #VALUE!).
  function serialOf(v) {
    if (v === null || v === undefined || v === '') return { kind: 'blank', serial: 0 };
    if (v instanceof Date) return { kind: 'num', serial: serialFromDate(v) };
    if (typeof v === 'number') return { kind: 'num', serial: v };
    if (typeof v === 'boolean') return { kind: 'num', serial: v ? 1 : 0 };
    return { kind: 'text', serial: null };
  }

  // Collapse an ExcelJS rich value to the plain value openpyxl would see.
  function plainVal(v) {
    if (v === null || v === undefined) return null;
    if (v instanceof Date) return v;
    if (typeof v === 'object') {
      if (v.richText) return v.richText.map(function (t) { return t.text; }).join('');
      if ('formula' in v || 'sharedFormula' in v) return v.result === undefined ? null : v.result;
      if ('text' in v) return v.text;   // hyperlink → display text
      if ('error' in v) return null;
      if ('result' in v) return v.result === undefined ? null : v.result;
      return null;
    }
    return v;
  }

  // Copy value preserving formulas ({formula, result}) like openpyxl copies
  // formula strings; everything else collapses to the plain value.
  function copyVal(v) {
    if (v !== null && v !== undefined && typeof v === 'object' && !(v instanceof Date) &&
        ('formula' in v || 'sharedFormula' in v)) {
      return { formula: v.formula || v.sharedFormula, result: v.result };
    }
    return plainVal(v);
  }

  // ── copy_sheet — coordinate-preserving value copy ───────────────────────
  async function copySheet(src, target, progress, label) {
    var rows = src.rowCount, cols = src.columnCount;
    for (var r = 1; r <= rows; r++) {
      var vals = src.getRow(r).values;   // sparse array indexed by column number
      for (var c = 1; c <= cols; c++) {
        var v = vals[c];
        if (v === undefined || v === null) continue;
        var nv = copyVal(v);
        if (nv === null) continue;
        var cell = target.getCell(r, c);
        cell.value = nv;
        // openpyxl auto-formats datetime cells; ExcelJS shows a raw serial
        // without a numFmt, so mirror openpyxl's default here.
        if (nv instanceof Date) cell.numFmt = 'yyyy-mm-dd h:mm:ss';
      }
      if ((r & 2047) === 0) {
        progress('Copying ' + label + ' — row ' + r.toLocaleString() + ' of ' + rows.toLocaleString() + '…');
        await _yield();
      }
    }
  }

  // ── find_column / add_column (the case-sensitivity asymmetry is Python's) ──
  function findColumn(ws, headerName) {
    var target = String(headerName).toLowerCase();
    for (var c = 1; c <= ws.columnCount; c++) {
      var v = plainVal(ws.getCell(1, c).value);
      if (v !== null && String(v).trim().toLowerCase() === target) return c;
    }
    return null;
  }
  function addColumn(ws, headerName) {
    for (var c = 1; c <= ws.columnCount; c++) {
      var v = plainVal(ws.getCell(1, c).value);
      if (v !== null && String(v).trim() === headerName) return c;
    }
    var newCol = ws.columnCount + 1;
    ws.getCell(1, newCol).value = headerName;
    return newCol;
  }

  async function convertDateColumn(ws, colNum) {
    if (!colNum) return;
    for (var r = 2; r <= ws.rowCount; r++) {
      var cell = ws.getCell(r, colNum);
      var v = cell.value;
      if (v === null || v === undefined) continue;
      if (v instanceof Date) { cell.numFmt = 'dd-mm-yyyy'; continue; }
      var parsed = parseDateString(String(v));
      if (parsed) { cell.value = parsed; cell.numFmt = 'dd-mm-yyyy'; }
      if ((r & 2047) === 0) await _yield();
    }
  }

  // ── process_risk_export ─────────────────────────────────────────────────
  async function processRiskExport(ws, warnings) {
    await convertDateColumn(ws, findColumn(ws, 'Date created'));
    var agingCol = addColumn(ws, 'Aging');
    var todaySerial = serialFromDate(todayUTC());
    for (var r = 2; r <= ws.rowCount; r++) {
      var n = serialOf(plainVal(ws.getCell(r, 14).value));  // column N — hardcoded like the Python
      var cellVal = { formula: 'NETWORKDAYS(N' + r + ',TODAY())' };
      if (n.kind !== 'text') cellVal.result = networkdaysSerial(n.serial, todaySerial);
      else warnings.push(SHEET_RISK + ' row ' + r + ': N is text — Aging left to Excel');
      ws.getCell(r, agingCol).value = cellVal;
      if ((r & 2047) === 0) await _yield();
    }
  }

  // ── process_assessment ──────────────────────────────────────────────────
  async function processAssessment(ws, warnings) {
    await convertDateColumn(ws, findColumn(ws, 'Date created'));
    await convertDateColumn(ws, findColumn(ws, 'Date submitted'));
    var ageingCol = addColumn(ws, 'Ageing');
    var working1Col = addColumn(ws, 'Working1');
    var working2Col = addColumn(ws, 'Working2');
    var sep26Col = addColumn(ws, '30sep26');
    var todaySerial = serialFromDate(todayUTC());
    var CUTOFF_2025 = serialFromDate(new Date(Date.UTC(2025, 11, 31))); // DATE(2025,12,31)

    for (var r = 2; r <= ws.rowCount; r++) {
      var cRaw = plainVal(ws.getCell(r, 3).value);   // column C
      var nRaw = plainVal(ws.getCell(r, 14).value);  // column N
      var oRaw = plainVal(ws.getCell(r, 15).value);  // column O
      var n = serialOf(nRaw), o = serialOf(oRaw);

      // Ageing
      var ageing = { formula: 'NETWORKDAYS(N' + r + ',TODAY())' };
      if (n.kind !== 'text') ageing.result = networkdaysSerial(n.serial, todaySerial);
      else warnings.push(SHEET_ASSESSMENT + ' row ' + r + ': N is text — Ageing left to Excel');
      ws.getCell(r, ageingCol).value = ageing;

      // Working1 — Excel semantics: text "=" is case-insensitive; text > number
      // ranks TRUE; blank compares as 0; blank cell equals "".
      var cTxt = (cRaw === null || cRaw === undefined) ? '' : String(cRaw).toLowerCase();
      var isDone = (cTxt === 'completed' || cTxt === 'under review');
      var oGT = (o.kind === 'num') ? o.serial > CUTOFF_2025
              : (o.kind === 'text');
      var oBlank = (o.kind === 'blank');
      ws.getCell(r, working1Col).value = {
        formula: 'IF(AND(OR(C' + r + '="Completed",C' + r + '="Under Review"),O' + r + '>DATE(2025,12,31)),' +
                 '"Completed in 2026",IF(O' + r + '="","Pending","Completed Before 2026"))',
        result: (isDone && oGT) ? 'Completed in 2026' : (oBlank ? 'Pending' : 'Completed Before 2026'),
      };

      // Working2
      var w2 = { formula: 'IF(DATE(2026,9,30)-N' + r + '>=365,"Beyond 1 Year Overdue","Current")' };
      if (n.kind !== 'text') w2.result = (SEP26_SERIAL - n.serial >= 365) ? 'Beyond 1 Year Overdue' : 'Current';
      else warnings.push(SHEET_ASSESSMENT + ' row ' + r + ': N is text — Working2 left to Excel');
      ws.getCell(r, working2Col).value = w2;

      // 30sep26
      var s26 = { formula: 'DATE(2026,9,30)-N' + r };
      if (n.kind !== 'text') s26.result = SEP26_SERIAL - n.serial;
      ws.getCell(r, sep26Col).value = s26;

      if ((r & 2047) === 0) await _yield();
    }
  }

  async function loadFirstSheet(file) {
    var wb = new ExcelJS.Workbook();
    await wb.xlsx.load(await file.arrayBuffer());
    var ws = wb.worksheets[0];   // Python's wb.active
    if (!ws) throw new Error('"' + file.name + '" has no worksheets');
    return ws;
  }

  // ── build — main(). files = { risk: File, assessment: File, tprm: File } ──
  async function build(files, onProgress) {
    var progress = onProgress || function () {};
    var warnings = [];
    if (!files.risk) throw new Error('Risk export file not found.');
    if (!files.assessment) throw new Error('Assessment export file not found.');
    if (!files.tprm) throw new Error('TPRM export file not found.');

    var out = new ExcelJS.Workbook();
    out.calcProperties.fullCalcOnLoad = true;   // Excel refreshes TODAY() maths on open
    var wsTprm = out.addWorksheet(SHEET_TPRM);
    var wsAssessment = out.addWorksheet(SHEET_ASSESSMENT);
    var wsRisk = out.addWorksheet(SHEET_RISK);

    progress('Reading risk export…');
    var srcRisk = await loadFirstSheet(files.risk);
    progress('Reading TPRM export…');
    var srcTprm = await loadFirstSheet(files.tprm);
    progress('Reading assessment export…');
    var srcAssessment = await loadFirstSheet(files.assessment);

    await copySheet(srcRisk, wsRisk, progress, 'risk export');
    await copySheet(srcTprm, wsTprm, progress, 'TPRM export');
    await copySheet(srcAssessment, wsAssessment, progress, 'assessment export');

    progress('Adding risk ageing formulas…');
    await processRiskExport(wsRisk, warnings);
    progress('Adding assessment formulas…');
    await processAssessment(wsAssessment, warnings);

    progress('Writing workbook…');
    var bytes = await out.xlsx.writeBuffer();
    return {
      bytes: bytes,
      warnings: warnings,
      stats: {
        risk: { rows: wsRisk.rowCount, cols: wsRisk.columnCount },
        tprm: { rows: wsTprm.rowCount, cols: wsTprm.columnCount },
        assessment: { rows: wsAssessment.rowCount, cols: wsAssessment.columnCount },
      },
    };
  }

  return {
    classify: classify,
    build: build,
    OUTPUT_NAME: OUTPUT_NAME,
    // exposed for tests
    _internal: {
      parseDateString: parseDateString,
      networkdaysSerial: networkdaysSerial,
      serialFromDate: serialFromDate,
      serialOf: serialOf,
      plainVal: plainVal,
    },
  };
})();

File: create_prp.py
import os
from datetime import datetime
from openpyxl import load_workbook, Workbook

# =====================================================
# PATHS
# =====================================================

BASE_DIR = os.path.dirname(os.path.abspath(__file__))

INPUT_FOLDER = os.path.join(BASE_DIR, "input")
OUTPUT_FOLDER = os.path.join(BASE_DIR, "output")

OUTPUT_FILE = "PRP Sample Jun (2).xlsx"

# =====================================================
# FIND FILES
# =====================================================

def find_source_files():

    risk_file = None
    assessment_file = None
    tprm_file = None

    print("\nScanning Input Folder...\n")

    if not os.path.exists(INPUT_FOLDER):
        raise Exception(
            f"Input folder not found:\n{INPUT_FOLDER}"
        )

    for file in os.listdir(INPUT_FOLDER):

        lower = file.lower()

        print(f"Found: {file}")

        if not lower.endswith(".xlsx"):
            continue

        if "risk-export" in lower:
            risk_file = os.path.join(INPUT_FOLDER, file)

        elif "assessment-export" in lower:
            assessment_file = os.path.join(INPUT_FOLDER, file)

        elif "tprm" in lower or "supplier" in lower:
            tprm_file = os.path.join(INPUT_FOLDER, file)

    print("\nDetected Files")
    print("-" * 50)
    print("Risk File       :", risk_file)
    print("Assessment File :", assessment_file)
    print("TPRM File       :", tprm_file)

    return risk_file, assessment_file, tprm_file


# =====================================================
# COPY SHEET
# =====================================================

def copy_sheet(source_ws, target_ws):

    for row in source_ws.iter_rows():

        for cell in row:

            target_ws[cell.coordinate] = cell.value


# =====================================================
# FIND COLUMN
# =====================================================

def find_column(ws, header_name):

    for col in range(1, ws.max_column + 1):

        value = ws.cell(1, col).value

        if value is not None:

            if str(value).strip().lower() == header_name.lower():

                return col

    return None


# =====================================================
# DATE CONVERTER
# =====================================================

def convert_date_column(ws, col_num):

    if not col_num:
        return

    formats = [
        "%d/%m/%Y",
        "%d-%m-%Y",
        "%m/%d/%Y",
        "%Y-%m-%d",
        "%m-%d-%Y"
    ]

    for row in range(2, ws.max_row + 1):

        cell = ws.cell(row=row, column=col_num)

        if cell.value is None:
            continue

        if isinstance(cell.value, datetime):
            cell.number_format = "dd-mm-yyyy"
            continue

        value = str(cell.value).strip()

        for fmt in formats:

            try:

                cell.value = datetime.strptime(
                    value,
                    fmt
                )

                cell.number_format = "dd-mm-yyyy"

                break

            except:
                pass


# =====================================================
# ADD COLUMN
# =====================================================

def add_column(ws, header_name):

    for col in range(1, ws.max_column + 1):

        value = ws.cell(1, col).value

        if value is not None:

            if str(value).strip() == header_name:

                return col

    new_col = ws.max_column + 1

    ws.cell(
        row=1,
        column=new_col
    ).value = header_name

    return new_col


# =====================================================
# RISK EXPORT
# =====================================================

def process_risk_export(ws):

    print("Processing Risk Export")

    date_created = find_column(
        ws,
        "Date created"
    )

    convert_date_column(
        ws,
        date_created
    )

    aging_col = add_column(
        ws,
        "Aging"
    )

    for row in range(2, ws.max_row + 1):

        ws.cell(
            row=row,
            column=aging_col
        ).value = (
            f"=NETWORKDAYS(N{row},TODAY())"
        )


# =====================================================
# ASSESSMENT
# =====================================================

def process_assessment(ws):

    print("Processing Assessment")

    date_created = find_column(
        ws,
        "Date created"
    )

    date_submitted = find_column(
        ws,
        "Date submitted"
    )

    convert_date_column(
        ws,
        date_created
    )

    convert_date_column(
        ws,
        date_submitted
    )

    ageing_col = add_column(
        ws,
        "Ageing"
    )

    working1_col = add_column(
        ws,
        "Working1"
    )

    working2_col = add_column(
        ws,
        "Working2"
    )

    sep26_col = add_column(
        ws,
        "30sep26"
    )

    for row in range(2, ws.max_row + 1):

        # Ageing

        ws.cell(
            row=row,
            column=ageing_col
        ).value = (
            f"=NETWORKDAYS(N{row},TODAY())"
        )

        # Working1

        ws.cell(
            row=row,
            column=working1_col
        ).value = (
            f'=IF(AND(OR(C{row}="Completed",'
            f'C{row}="Under Review"),'
            f'O{row}>DATE(2025,12,31)),'
            f'"Completed in 2026",'
            f'IF(O{row}="",'
            f'"Pending",'
            f'"Completed Before 2026"))'
        )

        # Working2

        ws.cell(
            row=row,
            column=working2_col
        ).value = (
            f'=IF(DATE(2026,9,30)-N{row}>=365,'
            f'"Beyond 1 Year Overdue",'
            f'"Current")'
        )

        # 30sep26

        ws.cell(
            row=row,
            column=sep26_col
        ).value = (
            f"=DATE(2026,9,30)-N{row}"
        )


# =====================================================
# MAIN
# =====================================================

def main():

    risk_file, assessment_file, tprm_file = (
        find_source_files()
    )

    if not risk_file:
        raise Exception(
            "Risk export file not found."
        )

    if not assessment_file:
        raise Exception(
            "Assessment export file not found."
        )

    if not tprm_file:
        raise Exception(
            "TPRM export file not found."
        )

    wb_new = Workbook()

    default_sheet = wb_new.active
    wb_new.remove(default_sheet)

    ws_risk = wb_new.create_sheet(
        "OneTrust - Risk Export"
    )

    ws_tprm = wb_new.create_sheet(
        "TPRM WebApp Portal"
    )

    ws_assessment = wb_new.create_sheet(
        "OneTrust Assessment"
    )

    print("\nLoading Risk File")
    wb_risk = load_workbook(risk_file)
    copy_sheet(
        wb_risk.active,
        ws_risk
    )

    print("Loading TPRM File")
    wb_tprm = load_workbook(tprm_file)
    copy_sheet(
        wb_tprm.active,
        ws_tprm
    )

    print("Loading Assessment File")
    wb_assessment = load_workbook(
        assessment_file
    )
    copy_sheet(
        wb_assessment.active,
        ws_assessment
    )

    process_risk_export(
        ws_risk
    )

    process_assessment(
        ws_assessment
    )

    os.makedirs(
        OUTPUT_FOLDER,
        exist_ok=True
    )

    output_path = os.path.join(
        OUTPUT_FOLDER,
        OUTPUT_FILE
    )

    wb_new.save(
        output_path
    )

    print("\nSUCCESS")
    print(f"\nOutput Created:\n{output_path}")


if __name__ == "__main__":
    main()